# SPY Next-Day Volatility Forecasting MLOps Service

**Author:** Praneshraj Tiruppur Nagarajan Dhyaneswar

**Date:** 07-10-2026

## Goal

To develop a reproducible ML workflow for forecasting **next-day SPY Parkinson variance** from daily OHLCV market data. The project is
designed to demonstrate production-minded ML engineering: leakage-safe feature construction, chronological evaluation, experiment tracking, model-artifact management, API validation and a planned standalone inference service.

The selected model is **HAR-style Ridge regression pipeline**

```text
StandardScaler → Ridge(alpha=1.0)
```

It uses 7 features available at the close of trading day t to forecast
Parkinson variance on the next trading day t+1:

- `har_d`, `har_w`, `har_m`: daily, 5-trading-day, and 22-trading-day Parkinson-variance components.
- `rv_lag1`, `abs_return`, `hl_range`, `log_volume`: supporting volatility, return, price-range, and liquidity-related features.

Feature timing is leakage-safe!! the target is shifted forward by one trading day and model performance is assessed using chronological train/validation/test splits. The **HAR-Ridge model is compared against a persistence baseline** and evaluated using four expanding-window walk-forward folds.

### Results

- **Validation RMSE:** 9.34% lower than the persistence baseline.
- **Final test RMSE:** 19.76% lower than the persistence baseline.
- **Walk-forward evaluation:** HAR-Ridge outperformed persistence in all four folds, with median RMSE improvement of 17.44%.
- **Experiment tracking:** MLflow records model parameters, metrics, evaluation artifacts, the serialized model and the validation-based promotion decision.
- **Inference prototype:** FastAPI `TestClient` checks verified `/health`, valid `/predict` inference and HTTP 422 rejection of invalid negative variance input.



**Note**: This is a volatility-forecasting and MLOps project, not a trading system. The companion GitHub repo will package the validated model artifact as a standalone FastAPI service with input-reference monitoring, pytest coverage, Docker packaging and GitHub Actions CI.

## 1. Install packages


In [1]:
# Packages mainly for MLflow experiment log and simple API packaging later.
# %pip -q install pandas numpy matplotlib scikit-learn mlflow joblib requests

%pip -q install mlflow joblib requests


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.2/50.2 kB 946.4 kB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.1/51.1 kB 1.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.3/44.3 kB 1.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 19.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.1/4.1 MB 17.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 14.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 268.7/268.7 kB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 148.8/148.8 kB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.9/114.9 kB 2.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 228.4/228.4 kB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 136.5/136.5 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 148.7/148.7 kB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.

## 2. Imports and project paths


In [2]:
# Standard-library modules used for paths, reproducibility, JSON and hidden input.
import io
import json
import os
import random
from datetime import datetime, timezone
from getpass import getpass
from pathlib import Path

# Numerical, plotting, HTTP, ML and experiment-tracking libraries.
import joblib
import matplotlib.pyplot as plt
import mlflow
import mlflow.sklearn
import numpy as np
import pandas as pd
import requests
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_absolute_error, mean_squared_error
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler


In [3]:
# Repeated run
SEED = 42
random.seed(SEED)
np.random.seed(SEED)

# Point to the Colab project folder.
PROJECT_ROOT = Path('/content/spy_rerun_project_3_06102026')
PROJECT_ROOT.mkdir(parents=True, exist_ok=True)
RAW_DATA_DIR = PROJECT_ROOT / 'data' / 'raw'
PROCESSED_DATA_DIR = PROJECT_ROOT / 'data' / 'processed'
REPORT_DIR = PROJECT_ROOT / 'reports' / 'figures'

# Output Folders
for directory in [RAW_DATA_DIR, PROCESSED_DATA_DIR, REPORT_DIR]:
    directory.mkdir(parents=True, exist_ok=True)

print('Project root:', PROJECT_ROOT)


Project root: /content/spy_rerun_project_3_06102026


## 3. Download and validate daily SPY data


In [4]:
## Setting the single instrument and requested historical range
SYMBOL = 'SPY'
START_DATE = '2000-01-01'
END_DATE = datetime.now(timezone.utc).date().isoformat()

# KEY!!
TWELVE_DATA_API_KEY = getpass('Twelve Data API key (hidden): ')

# Request daily OHLCV bars in CSV format from Twelve Data
response = requests.get(
    'https://api.twelvedata.com/time_series',
    params={
        'symbol': SYMBOL,
        'interval': '1day',
        'outputsize': 5000,
        'format': 'CSV',
        'apikey': TWELVE_DATA_API_KEY,
    },
    timeout=60,
)

# Raise an error for an unsuccessful HTTP response
response.raise_for_status()

# Fail early if the provider returned an API error instead of CSV market data
if response.text.lstrip().startswith('{'):
    raise RuntimeError(f'Twelve Data returned an error: {response.text[:500]}')


Twelve Data API key (hidden): ··········


In [5]:
# ## Setting the single instrument and requested historical range
# SYMBOL = 'SPY'
# START_DATE = '2000-01-01'
# END_DATE = datetime.now(timezone.utc).date().isoformat()

# # raw_df = pd.read_csv(io.StringIO(response.text), sep=';')

# raw_df = pd.read_csv("/content/spy_daily.csv")
# raw_df.head()

In [6]:
raw_df = pd.read_csv(io.StringIO(response.text), sep=';')
raw_df.columns = [column.strip().lower() for column in raw_df.columns]
print('Returned columns:', raw_df.columns.tolist())

rename_map = {
    'datetime': 'Date',
    'date': 'Date',
    'open': 'Open',
    'high': 'High',
    'low': 'Low',
    'close': 'Close',
    'volume': 'Volume',
}
raw_df = raw_df.rename(columns=rename_map)

#### Choose
required_cols = ['Date', 'Open', 'High', 'Low', 'Close', 'Volume']
missing_cols = [col for col in required_cols if col not in raw_df.columns]
if missing_cols:
    print(response.text[:1000])
    raise ValueError(f'Missing required columns: {missing_cols}')

# Order the rows from oldest to newest!!
raw_df['Date'] = pd.to_datetime(raw_df['Date'])
for column in ['Open', 'High', 'Low', 'Close', 'Volume']:
    raw_df[column] = pd.to_numeric(raw_df[column], errors='coerce')
raw_df = raw_df.sort_values('Date').dropna().reset_index(drop=True)

raw_path = RAW_DATA_DIR / 'spy_daily.csv'
raw_df.to_csv(raw_path, index=False)

print('Rows:', len(raw_df))
print('Date range:', raw_df['Date'].min().date(), 'to', raw_df['Date'].max().date())
display(raw_df.head())

Returned columns: ['datetime', 'open', 'high', 'low', 'close', 'volume']
Rows: 5000
Date range: 2006-11-20 to 2026-10-07


,Date,Open,High,Low,Close,Volume
0,2006-11-20,140.30000,140.74001,139.94000,140.50000,69174200
1,2006-11-21,140.49001,140.67000,140.28999,140.64000,51367900
2,2006-11-22,140.75000,141.16000,140.50000,140.92000,45505300
3,2006-11-24,140.24001,140.84000,140.20000,140.35001,30998000
4,2006-11-27,140.28000,140.35001,138.38000,138.42000,84545100


## 4. Build leakage-safe features


### Theory and Goal

Purpose: Forecast the **next day SPY volatility**... not the price direction or trading profit. Each feature is calculated using the info available by the close of day(t), **the target is the Prakinson varaince on day (t+1)**

### Parkinson variance
The daily volatility is approximated from the daily **high** and **low** prices

**σ̂ _P_t ^2 = [ln(H_t / L_t)]^2 / 4ln(2)**

H_t and L_t are the daily high and low prices. Obviously, al larger high-low range gives a higher variance estimate.

Note that **4ln(2)** comes from the original Parkinson estimator derivation..not a tuned model parameter. This estimator uses more intraday information than close-to-close returns, although it does not fully capture overnight price gaps.

### HAR-style features
Recent  volatility help predict future volatility. The HAR idea used daily, weekly and monthly volatility components

**RV_t+1 = ω + β_d RV_t (d) + β_w RV_t (w) + β_m RV_t (m) + ɛ_t+1**

Please note
- Daily variance: 1 trading day
- Weekly average variance: 5 trading days
- Monthly average variance: 22 trading days

SPY is US ETF, so 5 and 22 represent appx. US trading-week and trading-month windows, not calendar days.

### Feature timing

The target is shifted forward by one row:

**target_var(t) = σ̂ _P_t+1 ^2**

which means  row t uses only day t information to forecast day t+1. The first rows are dropped because rolling features need history. the last row is dropped because it has no next-day target.

In [7]:
# Using a few return lags and rolling windows for simple supporting analysis
FEATURE_WINDOWS = [1, 5, 10, 20]
ROLL_WINDOWS = [5, 10, 20]


def add_leakage_safe_features(df: pd.DataFrame):
    # Sort by date first and ensure every rolling feature follows the correct time order
    data = df.copy().sort_values('Date').reset_index(drop=True)

    # Daily close-to-close log return.
    data['log_return'] = np.log(data['Close'] / data['Close'].shift(1))

    # Simple intraday range relative to the closing price
    data['hl_range'] = (data['High'] - data['Low']) / data['Close']

    # Log volume helps reduce the large scale difference in volume values
    data['log_volume'] = np.log(data['Volume'])

    # Parkinson variance uses the daily high-low range as a volatility proxy.
    data['parkinson_var'] = (np.log(data['High'] / data['Low']) ** 2) / (4 * np.log(2))

    # HAR-style daily, weekly, and monthly volatility features .. check youtuber NEDL
    data['har_d'] = data['parkinson_var']
    data['har_w'] = data['parkinson_var'].rolling(5).mean()
    data['har_m'] = data['parkinson_var'].rolling(22).mean()

    # AddING the previous day's variance and the size of today's return
    data['rv_lag1'] = data['parkinson_var'].shift(1)
    data['abs_return'] = data['log_return'].abs()

    # Keep a few past returns for simple analysis..not for the final RV model
    for lag in FEATURE_WINDOWS:
        data[f'return_lag_{lag}'] = data['log_return'].shift(lag)

    # Add rolling return averages and volatility over short windows.
    for window in ROLL_WINDOWS:
        data[f'roll_mean_{window}'] = data['log_return'].rolling(window).mean()
        data[f'roll_vol_{window}'] = data['log_return'].rolling(window).std(ddof=0)

    # Shift forward so features from day t predict variance on day t+1
    data['target_var'] = data['parkinson_var'].shift(-1)
    data['target_date'] = data['Date'].shift(-1)

    # KeepING the final volatility feature set small and easy to explain
    rv_features = ['har_d', 'har_w', 'har_m', 'rv_lag1', 'abs_return', 'hl_range', 'log_volume']

    # DropPING early rows without enough history and the last row without a future target
    model_data = data.dropna(subset=rv_features + ['target_var', 'target_date']).reset_index(drop=True)
    return model_data, rv_features




In [8]:
##Building the final model dataframe and the chosen volatility feature list
model_df, RV_FEATURES = add_leakage_safe_features(raw_df)

# Every target date must be after its feature date.
assert model_df['target_date'].gt(model_df['Date']).all()

# Saiving the processed feature table for reproducibility
feature_path = PROCESSED_DATA_DIR / 'spy_features.csv'
model_df.to_csv(feature_path, index=False)

print('RV features:', RV_FEATURES)
display(model_df[['Date', 'target_date', 'parkinson_var', 'target_var']].head())
print('Forecast-origin range:', model_df['Date'].min().date(), 'to', model_df['Date'].max().date())

RV features: ['har_d', 'har_w', 'har_m', 'rv_lag1', 'abs_return', 'hl_range', 'log_volume']


,Date,target_date,parkinson_var,target_var
0,2006-12-20,2006-12-21,0.000008,0.000022
1,2006-12-21,2006-12-22,0.000022,0.000017
2,2006-12-22,2006-12-26,0.000017,0.000012
3,2006-12-26,2006-12-27,0.000012,0.000011
4,2006-12-27,2006-12-28,0.000011,0.000009


Forecast-origin range: 2006-12-20 to 2026-10-06


In [9]:
# Check random rows again using only data available on each forecast date
def test_no_future_information(df: pd.DataFrame, original: pd.DataFrame, checks: int = 25):
    rng = np.random.default_rng(SEED)
    sample_size = min(checks, len(df) - 31)
    indices = rng.choice(np.arange(30, len(df) - 1), size=sample_size, replace=False)

    for i in indices:
        row = df.iloc[i]

        # Recalculate the previous variance using data available up to this date.
        history = original.loc[original['Date'] <= row['Date']].copy()
        history_var = (np.log(history['High'] / history['Low']) ** 2) / (4 * np.log(2))
        expected_rv_lag1 = history_var.iloc[-2]

        # Recalculate the target variance from the stored next trading date
        target_row = original.loc[original['Date'] == row['target_date']].iloc[0]
        expected_target_var = (np.log(target_row['High'] / target_row['Low']) ** 2) / (4 * np.log(2))

        if not np.isclose(row['rv_lag1'], expected_rv_lag1, atol=1e-12):
            raise AssertionError(f'Previous variance leakage at {row.Date.date()}')

        if not np.isclose(row['target_var'], expected_target_var, atol=1e-12):
            raise AssertionError(f'Target alignment error at {row.Date.date()}')

    print('Leakage checks passed')


test_no_future_information(model_df, raw_df)

Leakage checks passed


## 5. Chronological split


In [10]:
# Splittign using target dates so each forecast belongs to the period being predicted
TRAIN_END = pd.Timestamp('2023-12-31')
VAL_END = pd.Timestamp('2025-12-31')

# Train on older targets, tune on later data, and keep 2026 for final testing
train_df = model_df.loc[model_df['target_date'] <= TRAIN_END].copy()
val_df = model_df.loc[(model_df['target_date'] > TRAIN_END) & (model_df['target_date'] <= VAL_END)].copy()
test_df = model_df.loc[model_df['target_date'] > VAL_END].copy()

print('Train rows:', len(train_df))
print('Validation rows:', len(val_df))
print('Test rows:', len(test_df))

Train rows: 4284
Validation rows: 502
Test rows: 192


## 6. Baseline and HAR-Ridge

Purpose: Try to **compare simple volatility baseline with HAR-style Ridge Model**

### Persistence Baseline
Assumption: Tomorrow's variance will be **same** as today's variance

RV(t+1) = RV(t)

This is an important benchmark since volatlity is persistent. A more complex model should only be kept if it **improves on this simple rule**.


### HAR-Ridge model

Uses daily, weekly and monthly Parkinson variance features together with a small number of supporting features. Ridge regression is used instead of plain OLS because several volatility features are correlated, especially the daily, weekly and monthly variance values.

The model is trained only on the training period. It is then used to predict validation and test rows without being refitted on future data.

Performance metrics:

- **MAE:** average absolute prediction error
- **RMSE:** square root of average squared error, which gives more weight to larger errors

Predicted variance cannot be negative, so negative model outputs are clipped to zero.




In [11]:
# MAE and RMSE in one small helper for every model comparison
def regression_metrics(y_true, y_pred):
    return {
        'mae': mean_absolute_error(y_true, y_pred),
        'rmse': mean_squared_error(y_true, y_pred) ** 0.5,
    }

# Persistence baseline predicts tomorrow's variance using today's variance
val_persist_pred = val_df['parkinson_var'].to_numpy()
test_persist_pred = test_df['parkinson_var'].to_numpy()

persist_val_metrics = regression_metrics(val_df['target_var'], val_persist_pred)
persist_test_metrics = regression_metrics(test_df['target_var'], test_persist_pred)

print('Validation persistence:', persist_val_metrics)
print('Test persistence:', persist_test_metrics)

Validation persistence: {'mae': 5.3368520251554e-05, 'rmse': 0.00021404008600812475}
Test persistence: {'mae': 3.6043831805850876e-05, 'rmse': 6.17805503674088e-05}


In [12]:
# Scale the features first.. then fit Ridge on the training period only
har_ridge = Pipeline([
    ('scaler', StandardScaler()),
    ('model', Ridge(alpha=1.0)),
])
har_ridge.fit(train_df[RV_FEATURES], train_df['target_var'])

# Predict later periods without fitting again on validation or test data
val_pred = np.clip(har_ridge.predict(val_df[RV_FEATURES]), 0, None)
test_pred = np.clip(har_ridge.predict(test_df[RV_FEATURES]), 0, None)

har_val_metrics = regression_metrics(val_df['target_var'], val_pred)
har_test_metrics = regression_metrics(test_df['target_var'], test_pred)

print('Validation HAR-Ridge:', har_val_metrics)
print('Test HAR-Ridge:', har_test_metrics)

Validation HAR-Ridge: {'mae': 4.942483578268321e-05, 'rmse': 0.00019404978526034157}
Test HAR-Ridge: {'mae': 2.921160152727888e-05, 'rmse': 4.9570910887896796e-05}


In [13]:
# Baseline and HAR-Ridge results into one table for clear comparison
comparison_df = pd.DataFrame([
    {'model': 'persistence', 'split': 'validation', 'n': len(val_df), **persist_val_metrics},
    {'model': 'persistence', 'split': 'test', 'n': len(test_df), **persist_test_metrics},
    {'model': 'har_ridge', 'split': 'validation', 'n': len(val_df), **har_val_metrics},
    {'model': 'har_ridge', 'split': 'test', 'n': len(test_df), **har_test_metrics},
])

comparison_path = PROCESSED_DATA_DIR / 'model_comparison_var.csv'
comparison_df.to_csv(comparison_path, index=False)

display(comparison_df)

,model,split,n,mae,rmse
0,persistence,validation,502,0.000053,0.000214
1,persistence,test,192,0.000036,0.000062
2,har_ridge,validation,502,0.000049,0.000194
3,har_ridge,test,192,0.000029,0.000050


## 7. Keeping the promotion logic simple and honest


In [14]:
val_improvement = 1.0 - (har_val_metrics['rmse'] / persist_val_metrics['rmse'])
promote_har = bool(har_val_metrics['rmse'] < 0.99 * persist_val_metrics['rmse'])

print('Persistence validation RMSE:', round(persist_val_metrics['rmse'], 8))
print('HAR-Ridge validation RMSE:', round(har_val_metrics['rmse'], 8))
print('Validation improvement:', round(val_improvement * 100, 2), '%')
print('Promote HAR-Ridge:', promote_har)


Persistence validation RMSE: 0.00021404
HAR-Ridge validation RMSE: 0.00019405
Validation improvement: 9.34 %
Promote HAR-Ridge: True


In [15]:
# Calculate percentage error reduction against the persistence baseline.
improvement_df = pd.DataFrame([
    {
        'split': 'validation',
        'mae_improvement_pct': 100 * (1 - har_val_metrics['mae'] / persist_val_metrics['mae']),
        'rmse_improvement_pct': 100 * (1 - har_val_metrics['rmse'] / persist_val_metrics['rmse']),
    },
    {
        'split': 'test',
        'mae_improvement_pct': 100 * (1 - har_test_metrics['mae'] / persist_test_metrics['mae']),
        'rmse_improvement_pct': 100 * (1 - har_test_metrics['rmse'] / persist_test_metrics['rmse']),
    },
])

improvement_df[['mae_improvement_pct', 'rmse_improvement_pct']] = (
    improvement_df[['mae_improvement_pct', 'rmse_improvement_pct']].round(2)
)

display(improvement_df)

,split,mae_improvement_pct,rmse_improvement_pct
0,validation,7.39,9.34
1,test,18.96,19.76


### Results Interpretation

Please note that persistence is a meaningful baseline because market volatility tends to persist over nearby trading days!

HAR-Ridge improved on persistence for both evaluation periods.

| Period | Persistence RMSE | HAR-Ridge RMSE | Improvement |
|---|---:|---:|---:|
| Validation | 0.000214 | 0.000194 | 9.3% lower |
| Test | 0.000062 | 0.000050 | 19.8% lower |

HAR-Ridge also improved MAE on both periods. This means that it reduced both the typical absolute forecast error and the larger errors captured more strongly by RMSE.

The test errors are lower than validation errors for both models. Maybe the final test period was likely calmer or easier to forecast than the validation period. The important result is that HAR-Ridge outperformed persistence on both periods without being refitted on validation data !!

### Why Ridge regression??

HAR features are related to each other.
- Eg: Daily variance is part of the weekly and monthly rolling averages.  This makes ordinary least squares coefficients unstable.

Ridge adds a small penalty to large coefficients.. to combat feature correlation and keep the model stable, interpretale and simple.

Feature scaling is needed before Ridge regression since the input features have very different numerical ranges. It ensures Ridge penalty apply more fairly across all features.

Thus, **HAR-Ridge is selected as the final volatility forecasting model**

## 8. Walk-forward evidence

Note than the earlier validation and test split gives one view of model performance. **A time period may be unusually calm or unusually volatile!!**

Walk-forward evaluation tests the model across several ordered historical periods. For each fold, the model is trained only on data available before that period and then predicts the next unseen period.

Purpose: To support stronger evidence that HAR-Ridge improves on persistence across different market conditions without using future information.


In [16]:
# Using expanding training windows and close to one-year unseen test periods
walk_forward_folds = [
    ('2022-08-31', '2022-09-01', '2023-08-31'),
    ('2023-08-31', '2023-09-01', '2024-08-31'),
    ('2024-08-31', '2024-09-01', '2025-08-31'),
    ('2025-08-31', '2025-09-01', '2026-09-10'),
]

# Keep only rows available before the final walk-forward test end date
walk_forward_data = model_df.loc[
    model_df['target_date'] <= pd.Timestamp('2026-09-10')
].copy()

print('Walk-forward rows:', len(walk_forward_data))

Walk-forward rows: 4959


In [17]:
walk_forward_results = []

for fold_number, (train_end, test_start, test_end) in enumerate(walk_forward_folds, start=1):

    # Training only on targets available before the current test period
    fold_train = walk_forward_data.loc[
        walk_forward_data['target_date'] <= pd.Timestamp(train_end)
    ].copy()

    # Testing on the next unseen period
    fold_test = walk_forward_data.loc[
        (walk_forward_data['target_date'] >= pd.Timestamp(test_start)) &
        (walk_forward_data['target_date'] <= pd.Timestamp(test_end))
    ].copy()

    # Fit new pipeline for this fold
    fold_model = Pipeline([
        ('scaler', StandardScaler()),
        ('model', Ridge(alpha=1.0)),
    ])
    fold_model.fit(fold_train[RV_FEATURES], fold_train['target_var'])

    # HAR-Ridge prediction with negative variance clipped to zero.. -ve varaince doesnt make sense
    fold_har_pred = np.clip(
        fold_model.predict(fold_test[RV_FEATURES]),
        0,
        None,
    )

    # Persistence--> current day's variance = tomorrow's prediction
    fold_persist_pred = fold_test['parkinson_var'].to_numpy()

    # Comparing both models using RMSE
    persist_rmse = regression_metrics(
        fold_test['target_var'],
        fold_persist_pred,
    )['rmse']

    har_rmse = regression_metrics(
        fold_test['target_var'],
        fold_har_pred,
    )['rmse']

    # Positive improvement means HAR-Ridge has lower RMSE than persistence!!
    har_improvement = 1 - (har_rmse / persist_rmse)

    walk_forward_results.append({
        'fold': fold_number,
        'train_target_end': fold_train['target_date'].max().date(),
        'test_target_start': fold_test['target_date'].min().date(),
        'test_target_end': fold_test['target_date'].max().date(),
        'train_rows': len(fold_train),
        'test_rows': len(fold_test),
        'persistence_rmse': persist_rmse,
        'har_ridge_rmse': har_rmse,
        'har_improvement_pct': har_improvement * 100,
    })

In [18]:
walk_forward_metrics = pd.DataFrame(walk_forward_results)

walk_forward_metrics['persistence_rmse'] = walk_forward_metrics['persistence_rmse'].round(8)
walk_forward_metrics['har_ridge_rmse'] = walk_forward_metrics['har_ridge_rmse'].round(8)
walk_forward_metrics['har_improvement_pct'] = walk_forward_metrics['har_improvement_pct'].round(2)

display(walk_forward_metrics)

median_improvement = walk_forward_metrics['har_improvement_pct'].median()

print('Median HAR-Ridge RMSE improvement:', round(median_improvement, 2), '%')
print('Positive median improvement:', median_improvement > 0)

,fold,train_target_end,test_target_start,test_target_end,train_rows,test_rows,persistence_rmse,har_ridge_rmse,har_improvement_pct
0,1,2022-08-31,2022-09-01,2023-08-31,3950,251,0.000125,0.000100,20.39
1,2,2023-08-31,2023-09-01,2024-08-30,4201,251,0.000044,0.000037,16.26
2,3,2024-08-30,2024-09-03,2025-08-29,4452,249,0.000298,0.000271,9.00
3,4,2025-08-29,2025-09-02,2026-09-10,4701,258,0.000070,0.000057,18.61


Median HAR-Ridge RMSE improvement: 17.44 %
Positive median improvement: True


### Walk-forward results interpretation

The median RMSE improvement was **17.44%**. In simple terms, the middle walk-forward result showed that HAR-Ridge reduced prediction error by about 17% compared with predicting tomorrow's variance as today's variance.

This supports the earlier validation result. It doesn't mean HAR-Ridge will win in every market period. It shows that the improvement was not limited to one specific split!!!

## 9. Save the selected model and feature reference
 HAR-Ridge passed validation promotion rule and with +ve walk forward results. Save this!

- Save the fitted `StandardScaler` and Ridge model together in one pipeline file
- This ensures future predictions use the same scaling learned from training data
- Save basic training-only statistics for every input feature: mean, std dev and percentiles
- These reference values can later help detect whether new input data is very different from the training data

In [19]:
if not promote_har:
    raise RuntimeError('HAR-Ridge did not pass the validation promotion rule')

print('HAR-Ridge passed the promotion rule and can be saved')

HAR-Ridge passed the promotion rule and can be saved


In [20]:
# Save the fitted scaler and Ridge model together in one file
model_path = PROCESSED_DATA_DIR / 'har_ridge_var.joblib'

joblib.dump(har_ridge, model_path)



['/content/spy_rerun_project_3_06102026/data/processed/har_ridge_var.joblib']

In [21]:
# Saving training-only feature values as a simple reference for later drift checks
reference_stats = {}

for column in RV_FEATURES:
    reference_stats[column] = {
        'mean': float(train_df[column].mean()),
        'std': float(train_df[column].std(ddof=0)),
        'q10': float(train_df[column].quantile(0.10)),
        'q50': float(train_df[column].quantile(0.50)),
        'q90': float(train_df[column].quantile(0.90)),
    }

reference_path = PROCESSED_DATA_DIR / 'har_feature_reference.json'
reference_path.write_text(json.dumps(reference_stats, indent=2))

print('Saved feature reference:', reference_path)
print('Feature Reference file exists:', reference_path.exists())
print('Saved model:', model_path)
print('Model file exists:', model_path.exists())

print('\nSaved files:')
for path in [model_path, reference_path]:
    print('-', path.name)

Saved feature reference: /content/spy_rerun_project_3_06102026/data/processed/har_feature_reference.json
Feature Reference file exists: True
Saved model: /content/spy_rerun_project_3_06102026/data/processed/har_ridge_var.joblib
Model file exists: True

Saved files:
- har_ridge_var.joblib
- har_feature_reference.json


## 10. MLflow experiment tracking

- Record model setup, including features, Ridge alpha and split dates
- Log baseline and HAR-Ridge evaluation metrics
- Save the model comparison CSV and selected trained pipeline as MLflow artifacts
- Record the final promotion decision for traceability


In [22]:
# Storing MLflow tracking data inside the project folder
mlflow_db_path = PROJECT_ROOT / 'mlflow.db'

os.environ['MLFLOW_TRACKING_URI'] = f'sqlite:///{mlflow_db_path}'
mlflow.set_tracking_uri(os.environ['MLFLOW_TRACKING_URI'])

mlflow.set_experiment('spy_volatility_forecast')

print('MLflow tracking URI:', mlflow.get_tracking_uri())

2026/10/07 16:19:07 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2026/10/07 16:19:07 INFO mlflow.store.db.utils: Updating database tables
2026/10/07 16:19:11 INFO mlflow.tracking.fluent: Experiment with name 'spy_volatility_forecast' does not exist. Creating a new experiment.


MLflow tracking URI: sqlite:////content/spy_rerun_project_3_06102026/mlflow.db


In [23]:
# Saving the walk-forward results so MLflow can log them as an artifact
walk_forward_path = PROCESSED_DATA_DIR / 'walk_forward_results.csv'
walk_forward_metrics.to_csv(walk_forward_path, index=False)

print('Saved walk-forward results:', walk_forward_path)
print('File exists:', walk_forward_path.exists())

Saved walk-forward results: /content/spy_rerun_project_3_06102026/data/processed/walk_forward_results.csv
File exists: True


In [24]:
# Record this training run, its results and the selected model artifact.
with mlflow.start_run(run_name='har_ridge_parkinson_variance') as run:
    # Save model settings and project choices
    mlflow.log_params({
        'model': 'StandardScaler + Ridge', 'target': 'next_day_parkinson_variance',
        'features': ', '.join(RV_FEATURES), 'train_end': str(TRAIN_END.date()),
        'validation_end': str(VAL_END.date()), 'ridge_alpha': 1.0,
    })

    # Saving evaluation metrics for both models
    mlflow.log_metrics({
        'persistence_val_rmse': float(persist_val_metrics['rmse']),
        'har_val_rmse': float(har_val_metrics['rmse']),
        'persistence_test_rmse': float(persist_test_metrics['rmse']),
        'har_test_rmse': float(har_test_metrics['rmse']),
        'validation_rmse_improvement': float(val_improvement),
        'walk_forward_median_improvement': float(median_improvement / 100),
    })

    # Labels to make comprehensive runsr
    mlflow.set_tags({
        'promotion_decision': str(promote_har), 'task_type': 'volatility_forecasting',
        'project_scope': 'not_a_trading_system',
    })

    # Useful output files
    mlflow.log_artifact(comparison_path)
    mlflow.log_artifact(walk_forward_path)
    mlflow.log_artifact(model_path)
    mlflow.log_artifact(reference_path)

    # SKlearn pipeline MLflow model format
    mlflow.sklearn.log_model(
        sk_model=har_ridge, name='har_ridge_model',
        input_example=train_df[RV_FEATURES].head(3),
    )

    run_id = run.info.run_id

print('MLflow run ID:', run_id)

MLflow run ID: 7eb41122022e4c23a59fe238dbb8f2d9


In [25]:
# Check that MLflow recorded the run
experiment = mlflow.get_experiment_by_name('spy_volatility_forecast')

runs = mlflow.search_runs(
    experiment_ids=[experiment.experiment_id]
)

display(runs[
    [
        'run_id',
        'params.model',
        'metrics.har_val_rmse',
        'metrics.har_test_rmse',
        'metrics.walk_forward_median_improvement',
        'tags.promotion_decision',
    ]
].head())

,run_id,params.model,metrics.har_val_rmse,metrics.har_test_rmse,metrics.walk_forward_median_improvement,tags.promotion_decision
0,7eb41122022e4c23a59fe238dbb8f2d9,StandardScaler + Ridge,0.000194,0.00005,0.17435,True


## 11. Final test-period plot

- Plot the actual next-day variance against persistence and HAR-Ridge predictions
- This is just for a visual sanity check alongside MAE and RMSE

In [27]:
import plotly.graph_objects as go

In [28]:
# Table to plot the final unseen test period
plot_df = test_df[['target_date', 'target_var']].copy()
plot_df['persistence_prediction'] = test_persist_pred
plot_df['har_ridge_prediction'] = test_pred

# Interactive line chart
fig = go.Figure()

fig.add_trace(
    go.Scatter(
        x=plot_df['target_date'],
        y=plot_df['target_var'],
        mode='lines',
        name='Actual variance'
    )
)

fig.add_trace(
    go.Scatter(
        x=plot_df['target_date'],
        y=plot_df['persistence_prediction'],
        mode='lines',
        name='Persistence'
    )
)

fig.add_trace(
    go.Scatter(
        x=plot_df['target_date'],
        y=plot_df['har_ridge_prediction'],
        mode='lines',
        name='HAR-Ridge'
    )
)

fig.update_layout(
    title='Next-Day Parkinson Variance: Final Test Period',
    xaxis_title='Target date',
    yaxis_title='Variance',
    template='plotly_white',
    hovermode='x unified'
)

# Add interactive zoom tools on the x-axis
fig.update_xaxes(
    rangeslider_visible=True,
    rangeselector=dict(
        buttons=[
            dict(count=1, label='1m', step='month', stepmode='backward'),
            dict(count=3, label='3m', step='month', stepmode='backward'),
            dict(count=6, label='6m', step='month', stepmode='backward'),
            dict(step='all', label='All')
        ]
    )
)

# Save the interactive chart
figure_path = REPORT_DIR / 'test_har_var_interactive.html'
fig.write_html(figure_path)

fig.show()

print('Saved figure:', figure_path)
print('Figure exists:', figure_path.exists())

Saved figure: /content/spy_rerun_project_3_06102026/reports/figures/test_har_var_interactive.html
Figure exists: True


## 12. FastAPI prediction service

- Load the saved HAR-Ridge pipeline once when the API starts
- Accept the three HAR variance features through JSON request
- Validate that input variance values are non-negative
- Return the predicted next-day Parkinson variance!
- Keep model training inside the notebook and use the API only for inference

In [29]:
from fastapi import FastAPI, HTTPException  ## API app
from pydantic import BaseModel, Field  ## BaseModel defines validated request/response schemas.. Field adds validation rules+metadata.

app = FastAPI(
    title='SPY Volatility Prediction API',   # API title
    description='Predicts next-day Parkinson variance using HAR-Ridge'   # API purpose
)

MODEL_PATH = model_path   # reusing saved notebook model path
model = joblib.load(MODEL_PATH)   # load saved scaler + Ridge pipeline


class PredictionInput(BaseModel):  ## Reminding again
    har_d: float = Field(ge=0)        # daily variance feature
    har_w: float = Field(ge=0)        # weekly variance feature
    har_m: float = Field(ge=0)        # monthly variance feature
    rv_lag1: float = Field(ge=0)      # previous day variance
    abs_return: float = Field(ge=0)   # absolute return size
    hl_range: float = Field(ge=0)     # intraday range scaled by close
    log_volume: float                 # log trading volume


@app.get('/health')
def health():
    return {'status': 'ok', 'model_loaded': True}   # basic API check


@app.post('/predict')
def predict(data: PredictionInput):
    features = pd.DataFrame([{   # same name as training
        'har_d': data.har_d,
        'har_w': data.har_w,
        'har_m': data.har_m,
        'rv_lag1': data.rv_lag1,
        'abs_return': data.abs_return,
        'hl_range': data.hl_range,
        'log_volume': data.log_volume,
    }])

    prediction = float(model.predict(features)[0])   # pipeline prediction
    prediction = max(prediction, 0.0)
    return {'predicted_variance': prediction}        # JSON response

### Quick local test

Use one sample row with the same training columns

In [30]:
test_payload = {
    'har_d': 0.00003,
    'har_w': 0.00004,
    'har_m': 0.00005,
    'rv_lag1': 0.00003,
    'abs_return': 0.01,
    'hl_range': 0.016,
    'log_volume': 18.0
}

print(test_payload)

{'har_d': 3e-05, 'har_w': 4e-05, 'har_m': 5e-05, 'rv_lag1': 3e-05, 'abs_return': 0.01, 'hl_range': 0.016, 'log_volume': 18.0}


In [31]:
from fastapi.testclient import TestClient   # local API testing

client = TestClient(app)   # attaching test client to API app

print(client.get('/health').json())   # test health endpoint
print(client.post('/predict', json=test_payload).json())   # test predict endpoint

{'status': 'ok', 'model_loaded': True}
{'predicted_variance': 3.296015390397444e-05}


### Result
- Saved HAR-Ridge pipeline loaded successfully.
- health endpoint confirmed the API was running.
- Predict endpoint accepted a valid JSON payload and returned next-day variance prediction

## Note: Dockerizing the FastAPI service

Purpose: To run the saved FastAPI inference service in a reproducible container outside the notebook

Why?? Docker makes the runtime portable and consistent across environments, which is more relevant to MLOps and DevOps than this notebook-only execution.

1. Move the API into a standalone Python file.
2. Save the trained model in a project `models/` folder.
3. Create `requirements.txt`.
4. Create a `Dockerfile` that starts Uvicorn.
5. Verify the container with `/health`.

Eventually, the docker image should build successfully and the running container should return HTTP 200 from the /health endpoint.

## 13. Prepare deployment files
notebook API prototype is useful for validation, but deployment needs fixed file paths
and reproducible runtime files outside notebook state


### Save model for standalone app

copy the trained pipeline into a normal project 'models/' folder so the standalone api and docker container loads it without depending on notebook variables

In [32]:
import shutil

In [33]:
APP_MODELS_DIR = Path('models')   # standalone app model folder
APP_MODELS_DIR.mkdir(exist_ok=True)

APP_MODEL_PATH = APP_MODELS_DIR / 'har_ridge_var.joblib'   # final app model path
shutil.copy(model_path, APP_MODEL_PATH)   # copy saved notebook model into app folder

print('saved app model =', APP_MODEL_PATH)
print('file exists =', APP_MODEL_PATH.exists())

saved app model = models/har_ridge_var.joblib
file exists = True


## 14. Testing API input validation

Prediction API **should reject invalid negative variance inputs** Verify that the Pydantic request schema protects the model from invalid requests before prediction.

In [34]:
invalid_payload = {
    'har_d': -0.00001,      # invalid: variance cannot be negative
    'har_w': 0.00004,
    'har_m': 0.00005,
    'rv_lag1': 0.00003,
    'abs_return': 0.01,
    'hl_range': 0.016,
    'log_volume': 18.0,
}

invalid_response = client.post('/predict', json=invalid_payload)

print('status code:', invalid_response.status_code)
print('response:', invalid_response.json())

assert invalid_response.status_code == 422

status code: 422
response: {'detail': [{'type': 'greater_than_equal', 'loc': ['body', 'har_d'], 'msg': 'Input should be greater than or equal to 0', 'input': -1e-05, 'ctx': {'ge': 0.0}}]}


## 15. Exporting validated model bundle for the GitHub repo

Exporting only the validated inference artifact and supporting evidence required by the standalone
repository service.

The export contains:

- The fitted `StandardScaler + Ridge` pipeline.
- Training-only reference statistics used for simple input-range monitoring.
- Model metadata, including the exact input-feature order and artifact checksum.
- Model-comparison and walk-forward evaluation tables.

No modification of the model done here!!!

In [35]:
import hashlib

In [36]:
required_variables = [
    "PROJECT_ROOT",
    "PROCESSED_DATA_DIR",
    "RV_FEATURES",
    "train_df",
    "val_df",
    "test_df",
    "comparison_df",
    "improvement_df",
    "walk_forward_metrics",
    "har_val_metrics",
    "har_test_metrics",
    "persist_val_metrics",
    "persist_test_metrics",
    "promote_har",
    "median_improvement",
    "run_id",
]

missing = [name for name in required_variables if name not in globals()]
if missing:
    raise RuntimeError(
        "Cannot export repository evidence; missing notebook state: "
        + ", ".join(missing)
    )

# Creating a repository-ready export folder without changing the original
# experiment artifacts stored under data/processed.
REPOSITORY_EXPORT_DIR = PROJECT_ROOT / "repository_export"
EXPORT_MODELS_DIR = REPOSITORY_EXPORT_DIR / "models"
EXPORT_REPORTS_DIR = REPOSITORY_EXPORT_DIR / "reports"

EXPORT_MODELS_DIR.mkdir(parents=True, exist_ok=True)
EXPORT_REPORTS_DIR.mkdir(parents=True, exist_ok=True)

# Use the fitted pipeline and training-only reference statistics created earlier
# The reference statistics will support simple production input-range warnings
SOURCE_MODEL_PATH = PROCESSED_DATA_DIR / "har_ridge_var.joblib"
SOURCE_REFERENCE_PATH = PROCESSED_DATA_DIR / "har_feature_reference.json"

if not SOURCE_MODEL_PATH.is_file():
    raise FileNotFoundError(f"Missing trained model: {SOURCE_MODEL_PATH}")

if not SOURCE_REFERENCE_PATH.is_file():
    raise FileNotFoundError(
        f"Missing training reference statistics: {SOURCE_REFERENCE_PATH}"
    )

# SHA-256 provides an immutable fingerprint of the serialized model
# Later, the API and README can prove which exact artifact was deployed
def sha256_file(path):
    return hashlib.sha256(path.read_bytes()).hexdigest()

EXPORTED_MODEL_PATH = EXPORT_MODELS_DIR / "har_ridge_var.joblib"
EXPORTED_REFERENCE_PATH = EXPORT_MODELS_DIR / "har_feature_reference.json"

# copy2 preserves file metadata where possible and leaves the original experiment artifacts untouched.
shutil.copy2(SOURCE_MODEL_PATH, EXPORTED_MODEL_PATH)
shutil.copy2(SOURCE_REFERENCE_PATH, EXPORTED_REFERENCE_PATH)

# Confirm that the repository copy is byte-for-byte identical to the validated model!
if sha256_file(SOURCE_MODEL_PATH) != sha256_file(EXPORTED_MODEL_PATH):
    raise RuntimeError("Model checksum mismatch after export.")

EXPORTED_COMPARISON_PATH = EXPORT_REPORTS_DIR / "model_comparison_var.csv"
EXPORTED_WALK_FORWARD_PATH = EXPORT_REPORTS_DIR / "walk_forward_results.csv"

# Export evaluation tables so repository claims can be checked without rerunning model training or trusting a README statement alone.
comparison_df.to_csv(EXPORTED_COMPARISON_PATH, index=False)
walk_forward_metrics.to_csv(EXPORTED_WALK_FORWARD_PATH, index=False)

# Record the train/validation/test periods and measured metrics in a compact machine-readable file for the README, tests, and future reproducibility checks.
metrics_summary = {
    "model_name": "har_ridge_var",
    "task": "next_day_parkinson_variance_forecasting",
    "instrument": "SPY",
    "feature_order": list(RV_FEATURES),
    "training_period": {
        "target_start": str(train_df["target_date"].min().date()),
        "target_end": str(train_df["target_date"].max().date()),
        "rows": int(len(train_df)),
    },
    "validation_period": {
        "target_start": str(val_df["target_date"].min().date()),
        "target_end": str(val_df["target_date"].max().date()),
        "rows": int(len(val_df)),
    },
    "test_period": {
        "target_start": str(test_df["target_date"].min().date()),
        "target_end": str(test_df["target_date"].max().date()),
        "rows": int(len(test_df)),
    },
    "metrics": {
        "persistence_validation": {
            key: float(value) for key, value in persist_val_metrics.items()
        },
        "har_ridge_validation": {
            key: float(value) for key, value in har_val_metrics.items()
        },
        "persistence_test": {
            key: float(value) for key, value in persist_test_metrics.items()
        },
        "har_ridge_test": {
            key: float(value) for key, value in har_test_metrics.items()
        },
        "validation_rmse_improvement_pct": float(
            improvement_df.loc[
                improvement_df["split"] == "validation",
                "rmse_improvement_pct",
            ].iloc[0]
        ),
        "test_rmse_improvement_pct": float(
            improvement_df.loc[
                improvement_df["split"] == "test",
                "rmse_improvement_pct",
            ].iloc[0]
        ),
        "walk_forward_median_rmse_improvement_pct": float(
            median_improvement
        ),
    },
    # This is a recorded validation-based model-selection decision .. not an automated production deployment decision.
    "promotion_decision": bool(promote_har),
    "mlflow_run_id": str(run_id),
}

EXPORTED_METRICS_PATH = EXPORT_REPORTS_DIR / "metrics_summary.json"
EXPORTED_METRICS_PATH.write_text(json.dumps(metrics_summary, indent=2))

# Store the inference contract next to the model. The standalone API will use this feature order and will apply the stated non-negative variance postprocessing.
model_metadata = {
    "artifact_name": "har_ridge_var.joblib",
    "artifact_sha256": sha256_file(EXPORTED_MODEL_PATH),
    "artifact_bytes": int(EXPORTED_MODEL_PATH.stat().st_size),
    "feature_order": list(RV_FEATURES),
    "model_pipeline": [
        "StandardScaler",
        "Ridge(alpha=1.0)",
    ],
    "prediction_postprocessing": (
        "The service returns max(raw_prediction, 0.0), because variance "
        "cannot be negative."
    ),
    "reference_file": "har_feature_reference.json",
    "reference_semantics": (
        "Training-only q10/q50/q90 statistics used for simple input "
        "reference-range warnings, not a formal drift test."
    ),
    "mlflow_run_id": str(run_id),
}

EXPORTED_METADATA_PATH = EXPORT_MODELS_DIR / "model_metadata.json"
EXPORTED_METADATA_PATH.write_text(json.dumps(model_metadata, indent=2))

# Confirm every required repository artifact exists and is non-empty before claiming that the export succeeded
exported_files = [
    EXPORTED_MODEL_PATH,
    EXPORTED_REFERENCE_PATH,
    EXPORTED_METADATA_PATH,
    EXPORTED_COMPARISON_PATH,
    EXPORTED_WALK_FORWARD_PATH,
    EXPORTED_METRICS_PATH,
]

for path in exported_files:
    if not path.is_file() or path.stat().st_size == 0:
        raise RuntimeError(f"Invalid exported file: {path}")

print("Repository export completed successfully:")
for path in exported_files:
    print(f"- {path.relative_to(PROJECT_ROOT)} ({path.stat().st_size} bytes)")

### The two most important artifact-contract checks for manual verification
print("\nModel SHA-256:", sha256_file(EXPORTED_MODEL_PATH))
print("Feature order:", list(RV_FEATURES))

Repository export completed successfully:
- repository_export/models/har_ridge_var.joblib (1561 bytes)
- repository_export/models/har_feature_reference.json (1317 bytes)
- repository_export/models/model_metadata.json (723 bytes)
- repository_export/reports/model_comparison_var.csv (292 bytes)
- repository_export/reports/walk_forward_results.csv (409 bytes)
- repository_export/reports/metrics_summary.json (1257 bytes)

Model SHA-256: 52da9c42b58671731f64f221dffed46f4aee4cf8c7828ff3c15a2b3ebd2a6ab4
Feature order: ['har_d', 'har_w', 'har_m', 'rv_lag1', 'abs_return', 'hl_range', 'log_volume']


## 16. Creating the Repo root

In [37]:
REPO_ROOT = Path("/content/spy_volatility_forecasting_mlops")
REPO_ROOT.mkdir(parents=True, exist_ok=True)

print("Repository root ready:", REPO_ROOT)
print("Exists:", REPO_ROOT.exists())

Repository root ready: /content/spy_volatility_forecasting_mlops
Exists: True


In [38]:
for folder in ["app", "models", "notebooks", "reports", "reports/figures", "tests", ".github/workflows"]:
    (REPO_ROOT / folder).mkdir(parents=True, exist_ok=True)

files_to_copy = [
    (REPOSITORY_EXPORT_DIR / "models" / "har_ridge_var.joblib",
     REPO_ROOT / "models" / "har_ridge_var.joblib"),

    (REPOSITORY_EXPORT_DIR / "models" / "har_feature_reference.json",
     REPO_ROOT / "models" / "har_feature_reference.json"),

    (REPOSITORY_EXPORT_DIR / "models" / "model_metadata.json",
     REPO_ROOT / "models" / "model_metadata.json"),

    (REPOSITORY_EXPORT_DIR / "reports" / "model_comparison_var.csv",
     REPO_ROOT / "reports" / "model_comparison_var.csv"),

    (REPOSITORY_EXPORT_DIR / "reports" / "walk_forward_results.csv",
     REPO_ROOT / "reports" / "walk_forward_results.csv"),

    (REPOSITORY_EXPORT_DIR / "reports" / "metrics_summary.json",
     REPO_ROOT / "reports" / "metrics_summary.json"),

    (REPORT_DIR / "test_har_var_interactive.html",
     REPO_ROOT / "reports" / "figures" / "test_har_var_interactive.html"),
]

for source, destination in files_to_copy:
    if not source.is_file():
        raise FileNotFoundError(f"Missing file: {source}")
    shutil.copy2(source, destination)

print("Repository files copied:")
for _, destination in files_to_copy:
    print("-", destination.relative_to(REPO_ROOT))

Repository files copied:
- models/har_ridge_var.joblib
- models/har_feature_reference.json
- models/model_metadata.json
- reports/model_comparison_var.csv
- reports/walk_forward_results.csv
- reports/metrics_summary.json
- reports/figures/test_har_var_interactive.html


## 17. Creating the first real repo source file

**app/schema.py**

In [39]:
%%writefile /content/spy_volatility_forecasting_mlops/app/schema.py
from pydantic import BaseModel, Field

FEATURE_ORDER = [
    "har_d",
    "har_w",
    "har_m",
    "rv_lag1",
    "abs_return",
    "hl_range",
    "log_volume",
]

class PredictionInput(BaseModel):
    har_d: float = Field(ge=0)
    har_w: float = Field(ge=0)
    har_m: float = Field(ge=0)
    rv_lag1: float = Field(ge=0)
    abs_return: float = Field(ge=0)
    hl_range: float = Field(ge=0)
    log_volume: float

Writing /content/spy_volatility_forecasting_mlops/app/schema.py


In [40]:
%%writefile /content/spy_volatility_forecasting_mlops/app/monitoring.py
import json
from pathlib import Path

from app.schema import FEATURE_ORDER


def load_reference(path: Path) -> dict:
    with path.open() as file:
        reference = json.load(file)

    if list(reference) != FEATURE_ORDER:
        raise ValueError("Reference feature order does not match model feature order.")

    return reference


def check_input_ranges(features: dict, reference: dict) -> dict:
    warnings = []

    for feature in FEATURE_ORDER:
        value = float(features[feature])
        q10 = float(reference[feature]["q10"])
        q90 = float(reference[feature]["q90"])

        if value < q10 or value > q90:
            warnings.append({
                "feature": feature,
                "value": value,
                "q10": q10,
                "q90": q90,
                "status": "outside_training_q10_q90",
            })

    return {
        "reference_type": "training_q10_q90_range_check",
        "warning_count": len(warnings),
        "warnings": warnings,
    }

Writing /content/spy_volatility_forecasting_mlops/app/monitoring.py


In [41]:
%%writefile /content/spy_volatility_forecasting_mlops/app/__init__.py
"""SPY volatility forecasting service package."""

Writing /content/spy_volatility_forecasting_mlops/app/__init__.py


In [42]:
import sys

if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))  # Make the repository package importable.

from app.schema import FEATURE_ORDER
from app.monitoring import check_input_ranges,load_reference

print("Import works.")
print(FEATURE_ORDER)

Import works.
['har_d', 'har_w', 'har_m', 'rv_lag1', 'abs_return', 'hl_range', 'log_volume']


In [43]:
reference_path = REPO_ROOT / "models" / "har_feature_reference.json"

reference = load_reference(reference_path)

normal_input = {
    feature: reference[feature]["q50"]
    for feature in FEATURE_ORDER
}

outside_range_input = normal_input.copy()
outside_range_input["har_d"] = reference["har_d"]["q90"] * 2

print("Normal input:")
print(check_input_ranges(normal_input, reference))

print("\nOutside-range input:")
print(check_input_ranges(outside_range_input, reference))

Normal input:
{'reference_type': 'training_q10_q90_range_check', 'warning_count': 0, 'warnings': []}

Outside-range input:
{'reference_type': 'training_q10_q90_range_check', 'warning_count': 1, 'warnings': [{'feature': 'har_d', 'value': 0.000407518981660161, 'q10': 7.5249173387426875e-06, 'q90': 0.0002037594908300805, 'status': 'outside_training_q10_q90'}]}


In [44]:
%%writefile /content/spy_volatility_forecasting_mlops/app/api.py
from pathlib import Path

import joblib
import pandas as pd
from fastapi import FastAPI

from app.monitoring import check_input_ranges, load_reference
from app.schema import FEATURE_ORDER, PredictionInput

ROOT_DIR = Path(__file__).resolve().parents[1]
MODEL_PATH = ROOT_DIR / "models" / "har_ridge_var.joblib"
REFERENCE_PATH = ROOT_DIR / "models" / "har_feature_reference.json"

model = joblib.load(MODEL_PATH)
reference = load_reference(REFERENCE_PATH)

app = FastAPI(
    title="SPY Next-Day Volatility Forecasting MLOps Service",
    description="Predicts next-day SPY Parkinson variance using HAR-Ridge.",
)


def feature_dict(request: PredictionInput) -> dict:
    return {
        feature: float(getattr(request, feature))
        for feature in FEATURE_ORDER
    }


@app.get("/health")
def health() -> dict:
    return {
        "status": "ok",
        "model_loaded": True,
        "model_file": MODEL_PATH.name,
    }


@app.post("/predict")
def predict(request: PredictionInput) -> dict:
    features = feature_dict(request)
    feature_frame = pd.DataFrame([features], columns=FEATURE_ORDER)

    raw_prediction = float(model.predict(feature_frame)[0])
    prediction = max(raw_prediction, 0.0)
    monitoring = check_input_ranges(features, reference)

    return {
        "predicted_variance": prediction,
        "monitoring": monitoring,
    }


@app.post("/monitor")
def monitor(request: PredictionInput) -> dict:
    return check_input_ranges(feature_dict(request), reference)

Writing /content/spy_volatility_forecasting_mlops/app/api.py


In [45]:
%%writefile /content/spy_volatility_forecasting_mlops/tests/test_api.py
from fastapi.testclient import TestClient

from app.api import app

client = TestClient(app)

VALID_PAYLOAD = {
    "har_d": 0.00003,
    "har_w": 0.00004,
    "har_m": 0.00005,
    "rv_lag1": 0.00003,
    "abs_return": 0.01,
    "hl_range": 0.016,
    "log_volume": 18.0,
}


def test_health_returns_ok():
    response = client.get("/health")

    assert response.status_code == 200
    assert response.json()["status"] == "ok"
    assert response.json()["model_loaded"] is True


def test_predict_returns_non_negative_variance():
    response = client.post("/predict", json=VALID_PAYLOAD)

    assert response.status_code == 200
    body = response.json()

    assert body["predicted_variance"] >= 0
    assert "monitoring" in body
    assert body["monitoring"]["warning_count"] == 0


def test_predict_rejects_negative_variance_input():
    invalid_payload = {**VALID_PAYLOAD, "har_d": -0.00001}

    response = client.post("/predict", json=invalid_payload)

    assert response.status_code == 422


def test_monitor_reports_outside_training_reference_range():
    outside_range_payload = {**VALID_PAYLOAD, "har_d": 0.0005}

    response = client.post("/monitor", json=outside_range_payload)

    assert response.status_code == 200
    body = response.json()

    assert body["warning_count"] >= 1
    assert any(
        warning["feature"] == "har_d"
        for warning in body["warnings"]
    )

Writing /content/spy_volatility_forecasting_mlops/tests/test_api.py


In [46]:
%%writefile /content/spy_volatility_forecasting_mlops/tests/conftest.py
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[1]

if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))

Writing /content/spy_volatility_forecasting_mlops/tests/conftest.py


## 18. Automated tests

In [47]:
%cd /content/spy_volatility_forecasting_mlops
!pytest -q

/content/spy_volatility_forecasting_mlops
....                                                                     [100%]
4 passed in 1.73s


In [48]:
%%writefile /content/spy_volatility_forecasting_mlops/requirements.txt
fastapi==0.115.6
uvicorn[standard]==0.32.1
pydantic==2.10.3
pandas==2.2.3
numpy==2.1.3
scikit-learn==1.6.1
joblib==1.6.0
httpx==0.28.1
pytest==8.3.4

Writing /content/spy_volatility_forecasting_mlops/requirements.txt


## 19. Creating Dockerfile

In [49]:
%%writefile /content/spy_volatility_forecasting_mlops/Dockerfile
FROM python:3.13-slim

WORKDIR /app

COPY requirements.txt .
RUN pip install --no-cache-dir -r requirements.txt

COPY app ./app
COPY models ./models

EXPOSE 8000

CMD ["uvicorn", "app.api:app", "--host", "0.0.0.0", "--port", "8000"]

Writing /content/spy_volatility_forecasting_mlops/Dockerfile


## 20. Creating .dockerignore

In [50]:
%%writefile /content/spy_volatility_forecasting_mlops/.dockerignore
.git
.github
.pytest_cache
__pycache__
*.py[cod]
.venv
venv
notebooks
reports
tests

Writing /content/spy_volatility_forecasting_mlops/.dockerignore


## 21. Create Github actions workflow.. Not running Docker in colab

In [51]:
%%writefile /content/spy_volatility_forecasting_mlops/.github/workflows/ci.yml
name: CI

on:
  push:
  pull_request:

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: Check out repository
        uses: actions/checkout@v4

      - name: Set up Python
        uses: actions/setup-python@v5
        with:
          python-version: "3.13"

      - name: Install dependencies
        run: pip install -r requirements.txt

      - name: Run tests
        run: pytest -q

      - name: Build Docker image
        run: docker build -t spy-volatility-mlops:ci .

Writing /content/spy_volatility_forecasting_mlops/.github/workflows/ci.yml


## 22. Create .gitignore

In [52]:
%%writefile /content/spy_volatility_forecasting_mlops/.gitignore
__pycache__/
*.py[cod]
.pytest_cache/
.venv/
venv/
.env
*.db
mlruns/
mlflow.db
data/raw/
data/processed/

Writing /content/spy_volatility_forecasting_mlops/.gitignore


### README

In [53]:
%%writefile /content/spy_volatility_forecasting_mlops/README.md
# SPY Next-Day Volatility Forecasting MLOps Service

This project forecasts **next-day SPY Parkinson variance** using daily OHLCV data.

The main goal was not to build a trading system. The purpose is to develop and
package a simple, leakage-safe volatility forecasting model with practical MLOps
components such as experiment tracking, model artifacts, API, testing, Docker
and CI.

> This project is not financial advice and I make no claims reg profitability

Daily SPY OHLCV data was obtained using the Twelve Data API. The API key is entered
interactively in the notebook and is not included in this repository.

## Project goal

The target is next-day Parkinson variance:

\[
\text{target\_var}(t) = \hat{\sigma}^{2}_{P,t+1}
\]

Features are calculated using information available by the close of trading day
\(t\), then used to forecast the next trading day \(t+1\).

The selected model is a scikit-learn pipeline--

```text
StandardScaler → Ridge(alpha=1.0)
```

## Features

The final HAR-style Ridge model uses 7 features:

- `har_d`: current-day Parkinson variance
- `har_w`: 5-trading-day average Parkinson variance
- `har_m`: 22-trading-day average Parkinson variance
- `rv_lag1`: previous-day Parkinson variance
- `abs_return`: absolute daily log return
- `hl_range`: high-low range scaled by close
- `log_volume`: log trading volume

The daily, weekly and monthly HAR-style features represent short, medium and
longer volatility history.

## Evaluation

Used chronological train, validation and test splits. I also compared the model
against a persistence baseline, where tomorrow's variance is predicted as today's
variance.

| Period | Persistence RMSE | HAR-Ridge RMSE | Improvement |
|---|---:|---:|---:|
| Validation | 0.000214 | 0.000194 | 9.34% lower |
| Final test | 0.000062 | 0.000050 | 19.76% lower |

I also used four expanding-window walk-forward folds. HAR-Ridge had lower RMSE than
persistence in all four folds, with a median improvement of **17.44%**.

The test-period visual comparison is available in:

```text
reports/figures/test_har_var_interactive.html
```

## MLOps components

- MLflow logs model parameters, metrics, artifacts and the promotion decision.
- The trained `StandardScaler + Ridge` pipeline is saved as a Joblib artifact.
- Model metadata includes feature order and a SHA-256 artifact checksum.
- FastAPI exposes `/health`, `/predict`, and `/monitor` endpoints.
- Pydantic rejects invalid negative variance-style inputs.
- `/monitor` compares request features against training-only q10-q90 reference
  ranges. This is a simple input-reference warning, not formal drift detection.
- Pytest covers health, valid prediction, invalid input and input monitoring.
- Docker packages the API and model artifact.
- GitHub Actions is configured to run tests and build the Docker image.

## Repository structure

```text
app/                 FastAPI service, schema, and monitoring logic
models/              Saved model, training reference, and metadata
notebooks/           Experiment and evaluation notebook
reports/             Evaluation tables and interactive test plot
tests/               Pytest API tests
.github/workflows/   CI workflow
Dockerfile           Container definition
requirements.txt     Pinned dependencies
```

## Run locally

```bash
python -m venv .venv
source .venv/bin/activate
pip install -r requirements.txt

pytest -q
uvicorn app.api:app --host 0.0.0.0 --port 8000
```

Health check:

```bash
curl -i http://127.0.0.1:8000/health
```

Example prediction:

```bash
curl -X POST http://127.0.0.1:8000/predict \
  -H "Content-Type: application/json" \
  -d '{
    "har_d": 0.00003,
    "har_w": 0.00004,
    "har_m": 0.00005,
    "rv_lag1": 0.00003,
    "abs_return": 0.01,
    "hl_range": 0.016,
    "log_volume": 18.0
  }'
```

## Docker

```bash
docker build -t spy-volatility-mlops:local .
docker run --rm -p 8000:8000 spy-volatility-mlops:local
```

Then call:

```bash
curl -i http://127.0.0.1:8000/health
```

## Current status

Verified in Colab:

- Repository FastAPI `/health`, `/predict`, and `/monitor` checks.
- Negative input rejection with HTTP 422.
- `pytest -q` passed with 4 tests.

Still to verify:

- Docker build and running-container health check.
- GitHub Actions workflow after the first push.
- Optional public cloud deployment.

## Limitations

- The API accepts precomputed features. It does not fetch live market data or create
  features from raw OHLCV requests.
- Parkinson variance does not fully capture overnight price gaps.
- This is a point-forecasting model. It does not provide uncertainty intervals.
- The input q10-q90 check is not a full drift-detection system.

Writing /content/spy_volatility_forecasting_mlops/README.md


In [54]:
print((REPO_ROOT / "README.md").exists())

True


In [56]:
%cd /content/spy_volatility_forecasting_mlops
!pytest -q

/content/spy_volatility_forecasting_mlops
....                                                                     [100%]
4 passed in 2.93s


In [57]:
%cd /content/spy_volatility_forecasting_mlops
!find . -type f | sort

/content/spy_volatility_forecasting_mlops
./app/api.py
./app/__init__.py
./app/monitoring.py
./app/__pycache__/api.cpython-313.pyc
./app/__pycache__/__init__.cpython-313.pyc
./app/__pycache__/monitoring.cpython-313.pyc
./app/__pycache__/schema.cpython-313.pyc
./app/schema.py
./Dockerfile
./.dockerignore
./.github/workflows/ci.yml
./.gitignore
./models/har_feature_reference.json
./models/har_ridge_var.joblib
./models/model_metadata.json
./.pytest_cache/CACHEDIR.TAG
./.pytest_cache/.gitignore
./.pytest_cache/README.md
./.pytest_cache/v/cache/nodeids
./README.md
./reports/figures/test_har_var_interactive.html
./reports/metrics_summary.json
./reports/model_comparison_var.csv
./reports/walk_forward_results.csv
./requirements.txt
./tests/conftest.py
./tests/__pycache__/conftest.cpython-313-pytest-8.4.2.pyc
./tests/__pycache__/test_api.cpython-313-pytest-8.4.2.pyc
./tests/test_api.py


## GIT

In [62]:
%cd /content/spy_volatility_forecasting_mlops

from getpass import getpass

GITHUB_USERNAME = "Jahprance"
GITHUB_REPO = "spy_volatility_forecasting_mlops"

token = getpass("Paste NEW token with repo + workflow scopes (hidden): ")

!git remote set-url origin https://{GITHUB_USERNAME}:{token}@github.com/{GITHUB_USERNAME}/{GITHUB_REPO}.git
!git push -u origin main

!git remote set-url origin https://github.com/{GITHUB_USERNAME}/{GITHUB_REPO}.git
del token

!git remote -v

/content/spy_volatility_forecasting_mlops
Paste NEW token with repo + workflow scopes (hidden): ··········
Enumerating objects: 28, done.
Counting objects: 100% (28/28), done.
Delta compression using up to 2 threads
Compressing objects: 100% (26/26), done.
Writing objects: 100% (28/28), 1.29 MiB | 1.96 MiB/s, done.
Total 28 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/Jahprance/spy_volatility_forecasting_mlops.git
 * [new branch]      main -> main
branch 'main' set up to track 'origin/main'.
origin	https://github.com/Jahprance/spy_volatility_forecasting_mlops.git (fetch)
origin	https://github.com/Jahprance/spy_volatility_forecasting_mlops.git (push)
